src/disc_bsd.c

```diff
@@ -0,0 +1,190 @@
+/* --------------------------------------------------------------------------
+
+   MusicBrainz -- The Internet music metadatabase
+
+   Copyright (C) 2008 Patrick Hurrelmann 
+   Copyright (C) 2006 Matthias Friedrich
+   Copyright (C) 2000 Robert Kaye
+   Copyright (C) 1999 Marc E E van Woerkom
+   
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+   
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+--------------------------------------------------------------------------- */
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/cdio.h>
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#if defined(__FreeBSD__)
+#include <netinet/in.h> /* for ntohl() */
+#else
+#include <util.h> /* for getrawpartition() */
+#endif
+
+#include "discid/discid_private.h"
+#include "unix.h"
+
+#define MAX_DEV_LEN 15
+
+static int get_device(int n, char* device_name, size_t device_name_length) {
+#if !defined(__FreeBSD__) /* /dev/rcdNX, where X is the letter for the raw partition */
+	snprintf(device_name, device_name_length, "/dev/rcd%d%c", n - 1, 'a' + getrawpartition());
+#else /* on FreeBSD it's just /dev/cdN */
+	snprintf(device_name, device_name_length, "/dev/cd%d", n - 1);
+#endif
+	return mb_disc_unix_exists(device_name);
+}
+
+int mb_disc_unix_read_toc_header(int fd, mb_disc_toc *toc) {
+	struct ioc_toc_header th;
+	struct cd_toc_entry te[100];
+	struct ioc_read_toc_entry rte;
+	int i;
+
+	memset(&th, 0, sizeof th);
+	if (ioctl(fd, CDIOREADTOCHEADER, &th) < 0)
+		return 0; /* error */
+
+	toc->first_track_num = th.starting_track;
+	toc->last_track_num  = th.ending_track;
+
+	if (toc->last_track_num == 0)
+		return 1; /* no entries to read */
+
```

Comment:
There should be a comment saying that we start reading the toc entries here.
If we do it like that.